Further Object.to_mesh() fixes for latest Blender 2.8 build

hedgehog90 opened the pull request:
Updated to work with latest Blender 2.8 API changes. I've made it backwards compatible for now so it still works with some earlier builds.
In the last few days Blender API has been changed so it now requires meshes created with 'Object.to_mesh()' to be removed with 'Object.to_mesh_clear()' instead of 'bpy.data.meshes.remove(mesh)'
Also fixed the annotation warning on startup, I made it a separate commit because I wasn't sure if keeping it like that was maybe intentional?

Jason0214 commented:
Thanks!

May I cherry-pick your commits to the #214? the `to_mesh_clear` is exactly what I miss in that PR.

Jason0214 commented on io_scene_godot/__init__.py:
```diff
@@ -46,11 +46,11 @@ class ExportGodot(bpy.types.Operator, ExportHelper):
 
     filename_ext = ".escn"
-    filter_glob = StringProperty(default="*.escn", options={"HIDDEN"})
+    filter_glob : StringProperty(default="*.escn", options={"HIDDEN"})
```
could you explain why this change?

hedgehog90 replied:
Blender 2.8 on loading produces warnings in the console if you don't set properties using 'annotations' (colons instead of equal signs), as in this:
https://github.com/KhronosGroup/glTF-Blender-IO/issues/147

hedgehog90 replied:
> Thanks!
> 
> May I cherry-pick your commits to the #214? the `to_mesh_clear` is exactly what I miss in that PR.

Sure. I've made further improvements since submitting the pull request though.
The 'Apply Modifiers' option was doing nothing, and no modifiers were being applied to meshes if it was checked.
Not all that familiar with github, if I push the new commit to my fork do I need to submit a new pull request?

Jason0214 replied:
You are doing correct, just push commits to your fork and github would update PR

Jason0214 commented:
wow, thank you for picking up my work :D.  Then let's merging this PR and close mine

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -67,6 +69,10 @@ def export_mesh_node(escn_file, export_settings, obj, parent_gd_node):
             escn_file, export_settings, mesh_node,
             obj.data.shape_keys, 'shapekey')
 
+    if ("ARMATURE" in export_settings['object_types'] and
```
I think it would be better to merge this block to the above `if ("ARMATURE"` block, which would seem to have more continuous logic. 

It can be merged writing as:
``` python
if AMRATURE_CHECK:
     # set to REST
     ...
     mesh_id = mesh_exporter.export_mesh(escn_file, export_settings)
     # set back
     ...
else:
     mesh_id = mesh_exporter.export_mesh(escn_file, export_settings)
```


Still, the current implementation is good enough, it depends on you to change it or not, I would merge it anyway :)

hedgehog90 replied:
Good points, will implement the specified changes.

Just curious if there's a good reason why we need to evaluate the object, to_mesh() it and triangulate it twice if it's a physics objects?
Why not do all that stuff at the top of export_mesh_node() ?

hedgehog90 replied:
Also, this:

`# Check if it exists so we don't bother to export it twice
        mesh = self.object.data
        mesh_id = escn_file.get_internal_resource(mesh)`

Doesn't take into account the mesh's modifiers. The same mesh data block can be used on multiple objects, but the modifiers could produce different meshes. This check would ignore the mesh objects with different modifers, if the mesh data block has already been processed.

Another reason to evaluate the object beforehand I think.

Jason0214 replied:
> Just curious if there's a good reason why we need to evaluate the object, to_mesh() it and triangulate it twice if it's a physics object?

Good question, it might be twice, but ideally it would be just once. First, in Blender there is no one to one mapping of Godot's CollisionShape node. In order to support it, the exporter export any Mesh node 
1. with `Rigid Body` 
2. draw type is `WIRE` 

as a physics node subtree(`Rigid` + `CollisionShape`).  Any Mesh satisfies both conditions break from the mesh node exporting [here](https://github.com/godotengine/godot-blender-exporter/blob/blender2.8/io_scene_godot/converters/mesh.py#L29). This is the idea way of creating a CollisionShape in Blender and mesh get exported once.

> Why not do all that stuff at the top of export_mesh_node(), send the resulting eval_obj to physics.export_physics_properties() and MeshResourceExporter()?

Good point, maybe we could make a abstract class namely `MeshNodeExporter` and inherits to a `PhysicalMeshExporter` and `MeshResourceExporter` so that some code could be shared.

Jason0214 replied:
>The same mesh data block can be used on multiple objects, but the modifiers could produce different meshes. This check would ignore the mesh objects with different modifers, if the mesh data block has already been processed.
> 

Good catch! This is definitely a bug, I will create an issue later and think about it more.

hedgehog90 replied:
I've tried to address some of the problems we've been discussing here, but it's not as neat as I'd hoped.
I've added a helper class - MeshConverter - which converts objects to meshes, applying modifiers (when specified), also handles triangulation and calculating tangents (when specified), reducing a bit of repetition across mesh.py and physics.py.
I've also re-arranged a few things and made some common function arguments in like escn_file and export_settings class variables in MeshResourceExporter because they're referred to so much.

However I'm holding off from pushing any changes until Blender fixes its API, there are bugs in all the recent blender builds that are making a fully functional exporter impossible right now.

Yesterday I realised evaluated_get on objects with modifiers was resulting in objects with empty vertex groups. This was fixed last night and the API was updated, however these new changes have resulted in a new bug, whereby Object.to_mesh() produces a mesh with no materials.

It's all a bit screwy right now, I wouldn't be surprised if they implement another overhaul on the whole depsgraph / to_mesh mess.

hedgehog90 replied:
Btw is there a problem if 2 resources in an escn have the same name?
As I mentioned before, with modifiers enabled it's possible you may use the same blender mesh more than once and create 2 unrelated meshes in godot, but the exporter uses mesh.name for mesh-related resource names which could result in names clashing. I've implemented a way to identify potential problem meshes and append them with some random alphanumeric string to prevent clashes, just wondering though if that's necessary?

Jason0214 replied:
> However I'm holding off from pushing any changes until Blender fixes its API, there are bugs in all the recent blender builds that are making a fully functional exporter impossible right now.

Yeah, just leave them until the APIs got stable. Thank you for your work and following the blender devs.

> I've added a helper class - MeshConverter - which converts objects to meshes, applying modifiers (when specified), also handles triangulation and calculating tangents (when specified), reducing a bit of repetition across mesh.py and physics.py.
I've also re-arranged a few things and made some common function arguments in like escn_file and export_settings class variables in MeshResourceExporter because they're referred to so much.

Wow, really appreciate it you doing this, it could be beneficial. I suggest you doing this refactoring in another PR, which would make the review and test easier

> Btw is there a problem if 2 resources in an escn have the same name?
No, actually the resource are directly using a blender object (in example it's `mesh`) as hash key, like
``` python
        mesh = self.object.data
        mesh_id = escn_file.get_internal_resource(mesh)
```
blender object is actually a C pointer, so name does not affect.

hedgehog90 replied:
> No, actually the resource are directly using a blender object (in example it's `mesh`) as hash key, like
> 
> ```python
>         mesh = self.object.data
>         mesh_id = escn_file.get_internal_resource(mesh)
> ```
> 
> blender object is actually a C pointer, so name does not affect.

I was referring to the resource_name property in a sub_resource.
I was wondering if this was ok:

```
[sub_resource id=1 type="ArrayMesh"]
resource_name = "Cube"
...
[sub_resource id=2 type="ArrayMesh"]
resource_name = "Cube"
...
```

I'm not sure, so I'm appending potential problem mesh resources with random alphanumeric strings just in case.

> I suggest you doing this refactoring in another PR, which would make the review and test easier

OK, I'll revert those particular changes for now as they're not significant, but there's a fair amount of difference between the last commit, but I've tried to keep mesh.py as similar as possible to the original.

Also, the Blender API bug where materials weren't carrying over after calling to_mesh() has just been fixed, so it should be released in tonight's beta release.

> Wow, really appreciate it you doing this, it could be beneficial.

And I really appreciate the amount of effort that you've put into this project. I haven't come across another blender exporter addon that manages to convert a 3D scene into another format so comprehensively. The material exporter in particular is excellent.
It's an invaluable tool for those of us working in Blender.

Jason0214 replied:
> I was referring to the resource_name property in a sub_resource.

Ah sorry for my misunderstanding. Yeah, you have a point. However, I did some test on, both Blender and Godot can resolve the resource name conflict, so we don't need to do anything about it.

> OK, I'll revert those particular changes for now as they're not significant, but although I've tried to keep mesh.py as similar as possible to the original, there's a fair amount of difference between my current version and the last commit.

You may not have to revert, you can use `git cherry-pick` to pick commits from one branch to another.

Thank you for your contribution as well ^ ^

Jason0214 replied:
Correction:

Yesterday I mentioned 
>Godot can resolve the resource name conflict

which may not be true, if the mesh import option set to `Files`, mesh resources with same name may overwrites each other and cause the error. However, blender seems has a strict checking against Mesh with same name, so this `resource_name` is still not important IMO

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -25,19 +26,23 @@ def export_mesh_node(escn_file, export_settings, obj, parent_gd_node):
+        # set armature to REST before exporting meshes
         armature_pose_position = armature_data.pose_position
         armature_data.pose_position = "REST"
+        # TODO: Not correct to assume the armature is closest parent. Armature
```
Great! You found this issue as well lol. It has been brought up in #194, I thought set a NodePath after exporting all nodes may not be enough.  Because in some case, the mesh needs to know information of its skeleton to be able to export vertex weights (e.g. when IK control bones are configured to not exported). 

IMO, we may need two traversals on blender scene tree. The first pass to built the node relations and the second pass export actual data.

hedgehog90 replied:
Agreed. But is there any other reason to do 2 traversals besides skeletons? Just curious.

Firstly, I suggest we add a bl_object_gd_node_map to ESCNFile, to which we add our NodeTemplate objects as values and the blender objects they're based on as keys (during export_object).

At the end of export_scene() we look through the exported blender objects, find the pairs of mesh objects and armatures that need to be linked, find the godot nodes with the map and set the node["skeleton"] links here.

Jason0214 replied:
By saying two passes,  I was mean to solve all the 'forward dependencies', not limit to Skeleton-Mesh. And I was thinking it would not be easy to fetch information of `exporting_bones` while exporting the mesh.  Well, by looking at your code, I found it's actually very easy to it :D 

Your approach is very good and certainly works, I would adopt it. Thanks

Jason0214 commented on io_scene_godot/export_godot.py:
```diff
@@ -136,7 +137,7 @@ def export_object(self, obj, parent_gd_node):
         if (exported_node.parent is not None and
                 exported_node.parent.get_type() == 'CollisionShape'):
             exported_node['transform'] = (
-                mathutils.Matrix.Rotation(math.radians(90), 4, 'X') @
```
This is not the same rotation (-90 vs 90),  a little hacky here.

hedgehog90 replied:
I changed this because its easier to interpret and it confused me at first, _AXIS_CORRECT is applied to the CollisionShape, so it makes sense _AXIS_CORRECT.invert() would be applied to the child.

Jason0214 replied:
Ah, I didn't notice the `invert`. Sorry, it's my fault. This is definitely a good change

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -155,63 +180,18 @@ def init_mesh_bones_data(self, skeleton_node):
-
+        """Saves a mesh into the escn file"""
+        mesh_converter = MeshConverter(self.object, export_settings)
+        key = (mesh_converter.resource_name, "Mesh")
```
do you think it's a good idea if we built a key from mesh resource and all the modifiers applied on it? (e.g. add a new class `MeshRscKey`. It's `__hash__()` method returns a hashable data which has same uniqueness as the exported mesh resource. This hashable data could be a serialization of `self.object.data` and all the modifiers name (seems it is compact).

Jason0214 replied:
`resource_name` which may built from a random combination of letters still has a possibility to have a collision

hedgehog90 replied:
Yes that makes a lot more sense.
I did consider making a hash key made up of (mesh, modifiers, modifier properties) but dismissed it as being too complicated, but it isn't really!

hedgehog90 commented:
Shall I commit further changes to this pull request or start another branch + pull request?
(apologies if I'm making a mess of the repo, I'm not overly familiar with git)

Jason0214 commented:
No bother to open a new one, just do it in this branch. They are all mesh related changes.

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -25,44 +24,49 @@ def export_mesh_node(escn_file, export_settings, obj, parent_gd_node):
 
-    mesh_node['mesh'] = "SubResource({})".format(mesh_id)
-    mesh_node['visible'] = obj.visible_get()
+    if armature_obj:
```
if "ARMATURE" not in export_settings['object_types', there is a risk of `armature_pose_position` being uninitialized

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -115,6 +116,78 @@ def export_object_link_material(escn_file, export_settings, mesh_object,
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
```
when comes to `__eq__`, I think it is usually not to compare the hash. Here you need to compare the original value (i.e. the tuple consists of objID and modifiers).  So that when different keys have a hash collision, we can still differentiate them with `__eq__`.

hedgehog90 replied:
I was only using __eq__ for testing, it's not used anywhere in the script (internally Dicts/hashtables will just use the __hash__ function, right?)
Anyway, considering it's a class with a single integer property it probably makes more sense to make MeshResourceKey a function for now.

Jason0214 replied:
actually `__eq__` is necessary for a hash key. When two keys has the same hash value, their selves need to be compare in addition to the `__hash__`, because `__hash__` always has possibility (although very small) to collision.

hedgehog90 replied:
Ah ok. Didn't know that. Fixing it.

Jason0214 commented on io_scene_godot/converters/__init__.py:
```diff
@@ -21,13 +21,17 @@
 from .animation import export_animation_data
 
-
+# TODO: What about Empties which refer to group instances? (in 2.8 Collection Instances)
```
You are right. There is also an issue requesting this addon to support `group instances`, but I haven't got chance to think about it 😅. Just leave a statement like `TODO: need  support for group instances` statement here, it can be dealt with later.

hedgehog90 replied:
There are a few ways we could do collection instances...
- convert the entire collection into meshes, join them into a single mesh (lossy)
- duplicate the collection's hierarchy under an empty spatial node (messy)
- export the collection as a separate escn, refer to it in the main escn as a PackedScene (many-files-y)

Regarding the first of those options - that's what I'd personally want to use for most of my group instances when exporting to Godot, but I can imagine other scenarios where I'd want to use the other 2 methods... So maybe we could allow for all 3 of these scenarios using object.name tags?

However I should note that providing several options for functionality that will rarely be used does seem a bit overkill.

Jason0214 commented on io_scene_godot/structures.py:
```diff
@@ -31,6 +31,7 @@ def __init__(self, heading):
         self.nodes = []
         self.internal_resources = []
         self._internal_hashes = {}
+        self._internal_resource_names = {}
```
is this attribute being used somewhere?

hedgehog90 replied:
Oops. No, removed now :)

Jason0214 commented on io_scene_godot/converters/__init__.py:
```diff
@@ -21,13 +21,17 @@
     "LIGHT": export_light_node,
-    "EMPTY": export_empty_node
+    "EMPTY": export_empty_node,
+    "CURVE": export_mesh_node,
```
Have you tried that all of these types node could be exported using the mesh exporting flow?

hedgehog90 replied:
I tried a few exports with each object type in a scene with modifiers and shapekeys applied (when possible) and I didn't notice any problems.

Jason0214 commented on io_scene_godot/export_godot.py:
```diff
@@ -93,19 +94,13 @@ def export_object(self, obj, parent_gd_node):
-        if (obj.type in converters.BLENDER_TYPE_TO_EXPORTER and
-                obj in self.exporting_objects):
+        if obj in self.exporting_objects:
             exporter = converters.BLENDER_TYPE_TO_EXPORTER[obj.type]
```
Looks like `BLENDER_TYPE_TO_EXPORTER` still can not cover all the object types. Maybe use `get` which is safer, instead of `[]`. And also keep the `Unknown object type` warning for debugging purpose

hedgehog90 replied:
Correct, reverted.

hedgehog90 commented:
It's not that weird - that blend file is saved with an object in edit mode and to_mesh won't work on an object currently in edit mode.
So I've added a few lines to export_scene to take it out of edit mode before export.

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -115,6 +116,79 @@ def export_object_link_material(escn_file, export_settings, mesh_object,
+class MeshResourceKey:
+    """Produces a key based on an object's data, its enabled modifiers
+    and their properties."""
+    # pylint: disable=protected-access
```
Better to move `# pylint: disable=protected-access` to function `__eq__` where it really affects. Smaller scope is better

Jason0214 commented on io_scene_godot/structures.py:
```diff
@@ -111,6 +112,7 @@ def __init__(self, entry_type, heading_dict=(), values_dict=()):
         self.entry_type = entry_type
         self.heading = collections.OrderedDict(heading_dict)
 
+        # NOTE: contents is unused. Remove?
```
I 'd rather keep it, if we really need some custom writing to node, it could be a good prompt

Jason0214 commented on io_scene_godot/structures.py:
```diff
@@ -132,21 +134,22 @@ def generate_heading_string(self):
-            self.generate_body_string(),
-            self.contents
-        )
+        heading = self.generate_heading_string()
```
I understand your changing would make the generated file more formatted. However I kind of believe an extra empty line does not matter, I would be much happier to make the code more simple and straightforward. So how about keep the original one?

Jason0214 replied:
Or maybe
```
        heading = self.generate_heading_string()
        body = self.generate_body_string()
        if body and self.contents:
            return "{}\n\n{}\n{}".format(heading, body, self.contents)
        if body:
            return "{}\n\n{}".format(heading, body)
        return heading
```

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -115,6 +116,79 @@ def export_object_link_material(escn_file, export_settings, mesh_object,
+    return (hasattr(object_data, "shape_keys") and
+            object_data.shape_keys is not None)
+
+def get_bl_rna_properties(obj):
```
This recursive traversal seems a little risky to me. I am thinking we may not need to do a very compact comparison (for example don't need to traversal on `bpy_prop_collection`), we can allow two identical mesh not being discovered by the inner_resource checking, because the only cost of this is an extra mesh exporting, which is acceptable.  For more information see the attached patch
[0001-Simplify-MeshResourceKey-construction.patch.zip](https://github.com/godotengine/godot-blender-exporter/files/3270100/0001-Simplify-MeshResourceKey-construction.patch.zip)

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -115,6 +116,79 @@ def export_object_link_material(escn_file, export_settings, mesh_object,
+    """Produces a key based on an object's data, its enabled modifiers
+    and their properties."""
+    # pylint: disable=protected-access
+    def __init__(self, obj, export_settings):
```
Here we may have to include the Godot resource name in the Key construction, because the same mesh can be used as ArrayMesh and CollisionShape at the same time. For more information, see the attached patch and the test case physics.escn
[0002-Include-Godot-Resource-Name-in-to-MeshResourceKey.patch.zip](https://github.com/godotengine/godot-blender-exporter/files/3270110/0002-Include-Godot-Resource-Name-in-to-MeshResourceKey.patch.zip)

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -249,44 +289,35 @@ def export_morphs(self, export_settings, surfaces):
-        blender_shape_keys = self.object.data.shape_keys
-        for index, shape_key in enumerate(blender_shape_keys.key_blocks):
-            if shape_key == blender_shape_keys.reference_key:
+        # Ran some tests, subsurface modifier and other modifiers appear to
```
I just did some test, yeah modifiers works fine with shape key now. Thanks for your effort! I think you can remove this part of code now.

Jason0214 commented on io_scene_godot/converters/mesh.py:
```diff
@@ -656,3 +685,70 @@ def __init__(self):
+        if mesh is not None:
+            triangulate_mesh(mesh)
+
+            # correct empty material bug in latest 2.8 release, remove on 30/05/19 build
```
Looks like this is not an issue anymore, remove this part?

Jason0214 commented:
Sorry for the late response,  got a little busy in last week.

I have reviewed and tested all the code. Overall looks very good to me and no regression except the MeshResourceKey one (test scene `physics.escn`) . If you fix the only issue and format the code, I will merge it. And here is also the test scene update [patch](https://github.com/godotengine/godot-blender-exporter/files/3270335/0001-Update-test-scene.patch.zip) if you need.
[001-Update-test-scene](https://github.com/godotengine/godot-blender-exporter/files/3270335/0001-Update-test-scene.patch.zip)

---
And looks like #194, #173, #217, #218 and #219 would be all fixed. Absolutely great work! Thanks you so much!

hedgehog90 commented:
I've applied the patches and changes suggested in this thread. Code is formatted and passes pycodestyle test.

I'm not sure what you meant about testing 'physics.escn', in the commit before last I added a few lines to take it out of editmode before export (obj to mesh conversion can't be done if the context is in editmode) which fixed the aforementioned error.

Although I updated the test folder escns with your '001-Update-test-scene' patch, it's still immediately failing on the first diff.

Jason0214 commented:
> I'm not sure what you meant about testing 'physics.escn'

Previously it is failing because we didn't add ResourceName to the KeyConstruction, now everything is good.

I am not sure if you can access this [link](https://travis-ci.org/godotengine/godot-blender-exporter/builds/544760730?utm_source=github_status&utm_medium=notification)
Travis is still complaining about the pylint formatting. Maybe you have a different verison of pylint (it changes frequently).

```
************* Module io_scene_godot
io_scene_godot/__init__.py:49:0: E0001: Variable annotation syntax is only supported in Python 3.6 and greater (<unknown>, line 49) (syntax-error)
************* Module io_scene_godot.export_godot
io_scene_godot/export_godot.py:35:0: E0001: Cannot import '' due to syntax error 'Variable annotation syntax is only supported in Python 3.6 and greater (<unknown>, line 49)' (syntax-error)
io_scene_godot/export_godot.py:36:0: E0001: Cannot import '' due to syntax error 'Variable annotation syntax is only supported in Python 3.6 and greater (<unknown>, line 49)' (syntax-error)
io_scene_godot/export_godot.py:175:4: R0912: Too many branches (14/12) (too-many-branches)
io_scene_godot/export_godot.py:31:0: W0611: Unused import math (unused-import)
io_scene_godot/export_godot.py:33:0: W0611: Unused import mathutils (unused-import)
************* Module io_scene_godot.converters.physics
io_scene_godot/converters/physics.py:8:0: W0611: Unused import bpy (unused-import)
io_scene_godot/converters/physics.py:10:0: W0611: Unused import bmesh (unused-import)
************* Module io_scene_godot.converters.mesh
io_scene_godot/converters/mesh.py:121:0: C0111: Missing function docstring (missing-docstring)
io_scene_godot/converters/mesh.py:132:0: C0111: Missing function docstring (missing-docstring)
io_scene_godot/converters/mesh.py:139:0: C0111: Missing function docstring (missing-docstring)
io_scene_godot/converters/mesh.py:144:0: C0111: Missing function docstring (missing-docstring)
io_scene_godot/converters/mesh.py:730:4: C0111: Missing method docstring (missing-docstring)
************* Module io_scene_godot.converters.material.script_shader.__init__
io_scene_godot/converters/material/script_shader/__init__.py:1:0: R0401: Cyclic import (io_scene_godot.converters.mesh -> io_scene_godot.converters.physics) (cyclic-import)
```
Mostly is about missing function docs. Feel free to put a `pylint: disable-msg` on some of these.

I will merge it after the formatting being fixed, and maybe squash the commits?

hedgehog90 commented:
Ah, I thought pylint wasn't an issue as there are a few unused imports before I started.
I take it I ignore the 'Variable annotation syntax' errors as Blender 2.8 uses Python 3.7.
After much faffing about, I managed to squash the lot into one commit (I think).
Hope that's OK.

Jason0214 commented:
Looks better now, the remaining warning from pylint is because travis is on python3.6
I will fixed myself on master.

Jason0214 commented:
Whoops, looks like the squashed commit does not linked to your github account, are you ok with it?

hedgehog90 commented:
> Whoops, looks like the squashed commit does not linked to your github account, are you ok with it?

Not a big deal but would it be possible to fix this?

Jason0214 commented:
Unfortunately I am not able to fix this one, it needs to *edit existing commit* and *force-push* the master branch. I am the Collaborator of this repo, it does not include *force-push* rights. 

I would love to see people who contributed can get some 'reward' for their effort. So @akien-mga could you able to editing the committer information of  [this](https://github.com/godotengine/godot-blender-exporter/commit/b534efb0aa4c13246fa45117a66c0144b00be7c5) and @hedgehog90 could you provide your email information registered at your github account?

----
To Remi:  I understand *force-push* would break all the existing forks, but probably I am the only on the only person active on this repo :P, so not a big deal

akien-mga commented:
There's no need to force push on the master branch, and I wouldn't do that just for changing the commit author email (unless it was actually wrong and impersonating someone).

@hedgehog90 You can simply add the email used in https://patch-diff.githubusercontent.com/raw/godotengine/godot-blender-exporter/pull/215.patch as secondary email in your GitHub settings, if you want to claim this identity as your own.

Jason0214 commented:
Thanks Remi, I didn't realize it can be changed in such a simple way.
